platform/measurement_utils.cpp

```diff
@@ -189,4 +192,97 @@ string FormatSpeed(double metersPerSecond)
   return res;
 }
 
+bool OSMDistanceToMeters(string const & osmRawValue, double & outMeters)
+{
+  char * stop;
+  char const * s = osmRawValue.c_str();
+  outMeters = strtod(s, &stop);
+
+  // Not a number, was not parsed at all.
+  if (s == stop)
+    return false;
+
+  if (!isfinite(outMeters))
+    return false;
+
+  switch (*stop)
+  {
+  // Default units - meters.
+  case 0: return true;
+
+  // Feet and probably inches.
+  case '\'':
+    {
+      outMeters = FeetToMeters(outMeters);
+      s = stop + 1;
+      double const inches = strtod(s, &stop);
+      if (s != stop && *stop == '"' && isfinite(inches))
+        outMeters += InchesToMeters(inches);
+      return true;
+    }
+    break;
+
+  // Inches.
+  case '\"': outMeters = InchesToMeters(outMeters); return true;
+
+  // It's probably a range. Use maximum value (if possible) for a range.
+  case '-':
+    {
+      s = stop + 1;
+      double const newValue = strtod(s, &stop);
+      if (s != stop && isfinite(newValue))
+        outMeters = newValue;
+    }
+    break;
+
+  // It's probably a list. Use maximum value (if possible) for a list.
+  case ';':
+    do
+    {
+      s = stop + 1;
+      double const newValue = strtod(s, &stop);
+      if (s == stop)
+        break;
+      if (isfinite(newValue))
+        outMeters = newValue;
+    } while (*stop);
+    break;
+  }
+
+  while (*stop && isspace(*stop))
```

Comment:
Пробелы съедаются strtod по документации.

> On 18.3.2016, at 15.13, ygorshenin <notifications@github.com> wrote:
> 
> In platform/measurement_utils.cpp:
> 
> > +
> > +  // It's probably a list. Use maximum value (if possible) for a list.
> > +  case ';':
> > +    do
> > +    {
> > +      s = stop + 1;
> > +      double const newValue = strtod(s, &stop);
> > +      if (s == stop)
> > +        break;
> > +      if (isfinite(newValue))
> > +        outMeters = newValue;
> > +    } while (*stop);
> > +    break;
> > +  }
> > +
> > +  while (*stop && isspace(*stop))
> Этот код тогда лучше и наверх - между числом и единицами измерения могут быть пробелы.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
>